Add projectile-comint-mode variable to allow for interactive compilation buffers

Toctave opened the pull request:
Hi,
This is my first PR here, I'm trying to follow the guidelines but I apologize in advance if I'm doing something wrong. I have not updated the readme or docs yet because I haven't found a suitable place to put my change and I didn't want to change the wrong files.

This is a very simple change that leverages Emacs' `compile` command argument to allow for interactive compilation buffers. This is particularly useful when using `projectile-project-run-cmd`, as it allows you to interact with your program.

-----------------

Before submitting a PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [x] The commits are consistent with our [contribution guidelines](../blob/master/CONTRIBUTING.md)
- [ ] You've added tests (if possible) to cover your change(s)
- [x] All tests are passing ([`eldev test`](https://github.com/doublep/eldev))
- [x] The new code is not generating bytecode or `M-x checkdoc` warnings
- [x] You've updated the [changelog](../blob/master/CHANGELOG.md) (if adding/changing user-visible functionality)
- [x] You've updated the readme (if adding/changing user-visible functionality)

Thanks!

bbatsov commented on projectile.el:
```diff
@@ -4447,11 +4447,14 @@ project of that type"
       (projectile-read-command prompt default-cmd)
     default-cmd))
 
+(defvar projectile-comint-mode nil
```
Shouldn't this be a `defcustom`?

Toctave replied:
It should indeed. I just fixed it

bbatsov commented:
It's good to mention this somewhere in the docs, otherwise probably no one will find it.

Toctave commented:
Just added a section at the end of [Configuring projectile's behavior](https://docs.projectile.mx/projectile/2.2/projects.html#configuring-projectiles-behavior) . Hope this was the right place to do so?

Toctave commented:
I just noticed this was requested in #1669 . My approach is different from the one suggested in that thread. Any thoughts @bbatsov ?

bbatsov commented on projectile.el:
```diff
@@ -4447,11 +4447,15 @@ project of that type"
       (projectile-read-command prompt default-cmd)
     default-cmd))
 
+(defcustom projectile-comint-mode nil
```
Probably this should be named `projectile-use-compilation-shell-minor-mode` or something along those lines, as `comint-mode` is something else and that would be confusing IMO.

bbatsov replied:
Oh, well the description is actually a bit confusing and I don't have time to test this myself:

> If optional second arg COMINT is t the buffer will be in Comint mode with
‘compilation-shell-minor-mode’.

Toctave replied:
It is both : when a prefix argument is used, the major mode of the compilation buffer is `comint-mode`, and the minor mode `compilation-shell-minor-mode` is activated. `comint-mode` is the more important one, though. All `compilation-shell-minor-mode` does is rebind keys to avoid conflicts : 

```
When Compilation Shell minor mode is enabled, all the
error-parsing commands of the Compilation major mode are
available but bound to keys that don't collide with Shell mode.
```

Perhaps `projectile-comint` would be the best choice of name : it uses `comint`, which is the name of the argument of the `compile` function and of the major mode the compilation buffer will be in, and avoids the `-mode` prefix, which might wrongly imply that this is an actual mode. Would you be fine with that ?

Toctave replied:
For completeness's sake, here are the bindings defined in `compilation-shell-minor-mode-map` :

```
C-M-n  compilation-next-error
C-M-p  compilation-previous-error
M-RET  compile-goto-error
M-{    compilation-previous-file
M-}    compilation-next-file
```

bbatsov replied:
Got it. Okay, let's name this `projectile-compilation-use-comint-mode` or something like this. I just want to make it a bit more explicit that this is related to complition commands.

bbatsov commented on doc/modules/ROOT/pages/projects.adoc:
```diff
@@ -665,6 +665,17 @@ your project, you could customize it with the following:
+By default, compilation buffers are not writable, which allows you to
+e.g.  press `g` to restart the last command. Setting
+`projectile-comint-mode` to a non-nil value allows you to make
+projectile compilation buffers interactive, letting you e.g. test a
```
I don't think "interactive" is the write term here.

Toctave replied:
I used that word because it is the one that is used in the Emacs documentation of `compile` (`C-h f compile <ENTER>`): 

```
Additionally, with universal prefix arg, compilation buffer will be in
comint mode, i.e. interactive.
```

bbatsov replied:
Got it.

bbatsov commented:
The two approaches are not mutually exclusive. I guess we can still have some prefix argument for the cases when people want this functionality on demand.

bbatsov commented on projectile.el:
```diff
@@ -4447,11 +4447,15 @@ project of that type"
+  "If non-nil, launch projectile compilation buffers in interactive mode."
+  :type 'bool)
+
 (defun projectile-run-compilation (cmd)
```
I think you can also add here an optional argument called `comint-mode`, so that people can triggered potentially with a prefix arg from interactive commands.

Toctave replied:
I would, but the user level function is `projectile-compile-project`, which already has a prefix argument (used to force the command prompt). Should I remove it in favour of this one ?

bbatsov commented:
You'll also need to rebase.

Toctave commented:
I've rebased and added a variable for each type of compilation command (didn't want to refactor the `projectile-<cmd>-project` code, but it's quite repetitive). 

Only thing left is to (maybe?) change the use of the prefix argument for the `projectile-<cmd>-project` functions.

bbatsov commented on doc/modules/ROOT/pages/projects.adoc:
```diff
@@ -715,6 +715,18 @@ your project, you could customize it with the following:
+
+[source,elisp]
+----
+(setq projectile-comint-mode t)
```
You forgot to update this.

Toctave replied:
What do you mean ? I did update it to reflect the new variable names

bbatsov commented:
> I've rebased and added a variable for each type of compilation command (didn't want to change the structure of the projectile-<cmd>-project code, but it's quite repetitive)

I meant this in the sense that all those commands use `compile` internally. :-) Probably we're fine with a single config for all commands.

bbatsov commented on projectile.el:
```diff
@@ -4536,11 +4536,11 @@ project of that type"
     default-cmd))
 
-(defun projectile-run-compilation (cmd)
+(defun projectile-run-compilation (cmd use-comint-mode)
```
I'd make `use-comint-mode` optional (`&optional).

bbatsov commented on projectile.el:
```diff
@@ -4582,9 +4582,39 @@ The command actually run is returned."
+(defcustom projectile-configure-use-comint-mode nil
+  "If non-nil, make the output buffer of projectile-configure-project
+interactive."
+  :type 'bool)
```
There should be a few other properties here. Refer to existing `defcustoms` for those.

Toctave commented:
> > I've rebased and added a variable for each type of compilation command (didn't want to change the structure of the projectile--project code, but it's quite repetitive)
> 
> I meant this in the sense that all those commands use `compile` internally. :-) Probably we're fine with a single config for all commands.

I'm not sure : my use case would be to make the `run` buffer interactive, and keep all other ones in compilation mode. This means I keep the syntax highlighting and links to errors in the compilation buffer, and I can interact with my program when I test it out.

bbatsov commented:
Got it. Well, I guess we can go with the granular defcustoms then.
